openvdb_points/tools/AttributeGroup.h

```diff
@@ -0,0 +1,226 @@
+//
+// All rights reserved. This software is distributed under the
+// Mozilla Public License 2.0 ( http://www.mozilla.org/MPL/2.0/ )
+//
+// Redistributions of source code must retain the above copyright
+// and license notice and the following restrictions and disclaimer.
+//
+// *     Neither the name of Double Negative Visual Effects nor the names
+// of its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file AttributeGroup.h
+///
+/// @authors Dan Bailey
+///
+/// @brief  Attribute Group access and filtering for iteration.
+///
+
+
+#ifndef OPENVDB_TOOLS_ATTRIBUTE_GROUP_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_ATTRIBUTE_GROUP_HAS_BEEN_INCLUDED
+
+#include <openvdb/openvdb.h>
+#include <openvdb_points/tools/AttributeArray.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+typedef uint8_t GroupType;
+
+
+////////////////////////////////////////
+
+
+class GroupAttributeArray : public TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >
+{
+public:
+    /// Default constructor, always constructs a uniform attribute.
+    explicit GroupAttributeArray(   size_t n = 1,
+                                    const ValueType& uniformValue = zeroVal<ValueType>());
+    /// Deep copy constructor (optionally decompress during copy).
+    GroupAttributeArray(const TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >& array,
+                        const bool decompress = false);
+
+    /// Cast an AttributeArray to GroupAttributeArray
+    static GroupAttributeArray& cast(AttributeArray& attributeArray);
+
+    /// Cast an AttributeArray to GroupAttributeArray
+    static const GroupAttributeArray& cast(const AttributeArray& attributeArray);
+
+    /// Return @c true if the AttributeArray provided is a group
+    static bool isGroup(const AttributeArray& attributeArray);
+
```

Comment:
Could isGroup live on the AttributeArray? It seems to be returning GroupAttributeArray::cast(attributeArray).isGroup() which is why it can't, and I'm wondering why .isType<GroupAttributeArray>() isn't enough to infer. My understanding currently is that if a GroupAttributeArray exists, then it must be a group but maybe that's not always the case?